### PR TITLE
[Request for Contributions (Issues)] Adding test to show that fit() is reducing reproducibility.

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
Hi!

Folowing #8906 

Calling fit changes the numpy random state. I don't think this is a bug, but I don't have much experience so I'm giving it the benefit of the doubt. Please tell me if this necessitate changes in the codebase, so that I can make a PR.

Citing @MaxG87 :

> Checking the source code I noticed that model.fit finally invokes np.random.shuffle(...), thus altering the global random state. This can be fixed very easily by introducing a new member of type np.random.RandomState, and then using this one instead of the global one.
I would appreciate very much if this could be fixed as it would remove one source of irreproducibility.
In my opinion, calling functions should not alter the global random state.

### CODE REVIEW

**Comment by MaxG87** [comment]
I was asked to explain my point of view over at #8906.

I think the described behaviour is very undesirable. It does not only alter the behaviour of other functions / objects depending on the numpy random state. It also introduces difficulties when pausing and continuing the training of models.

I found this when I tried to implement my own variant of HyperBand random search. There I train some models for some epochs, drop the worst ones and continue to train the best ones. However, there the order in which each element sees the data depends on how many models were trained how long before. This totally hurts reproducibility. Given some set of hyperparameters, the performance cannot be reproduced, as the order of the samples cannot be reproduced.

So, fixing this will be an important step towards easier reproducibility of model training. Not to infer with other codes randomness then only comes as a bonus.

**Comment by taehoonlee** [comment]
I think it is natural of `fit()` to consume the global random state. Otherwise, reproducibility may decrease because only `fit()` will be affected by another random state.

**Comment by gabrieldemarmiesse** [comment]
That is what I though too. Closing the PR now.

**Comment by gabrieldemarmiesse** [comment]
Thanks @taehoonlee for reading it. 

to @MaxG87 you can use this wrapper to solve your issues on your end, even if keras doesn't do it automatically:

```
import numpy as np

def no_grs(func):
    def new_function(*args):
        state = np.random.get_state()
        result = func(*args)
        np.random.set_state(state)
        return result
    return new_function

def fit(dummy):
    # Consume grs
    np.random.rand()
    return dummy + " and other dummy"

np.random.seed(15)
ref_val = np.random.rand()

np.random.seed(15)
my_value = no_grs(fit)("Dummy")
new_ref = np.random.rand()

print(my_value)
assert ref_val == new_ref
```

result:

`Dummy and other dummy
`

**Comment by MaxG87** [comment]
I accept that this is settled as won't fix, but for the unfortunate landing here through search engines I want to point out a last point. I ask for your understanding.

The above wrapper might improve reproducibility, but it might harm training. In my programs I don't consume any global randomness, so if I were about to use this wrapper, my models would be trained with the same permutation of training data all the time.

I ended up writing my own wrappers for data and models, only using train_on_batch and shuffling manually.

**Comment by kavinvin** [comment]
While this pr was already closed, I want to support the statement that using the global random state is very undesirable for reproducibility. Actually, I strongly believe that every function shouldn't use the global random state except for convenience.

I suggest that `model.fit` should create there own `np.random.RandomState` and give users absolute control to select their own seed. There is no reason to force a user to use the mutable global random state to make already unpredictable training more unpredictable. If randomness is actually needed, just don't use seed.

I'm now using my own generator to make the results consistent. But I think it's still a great improvement.
